Add check_dict method to Python grader

harivamsi9 opened the pull request:
Can you please review this PR and let me know the feedback?

This PR will resolve #8831

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,59 @@ def bad(msg):
+                return False
+
+        if data is None:
+            return bad("'%s' is None or not defined" % name)
```
```suggestion
            return bad(f"{name} is None or not defined")
```
Throughout the new changes, should use f-strings instead of the `%`.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,59 @@ def bad(msg):
+            return bad("'%s' is None or not defined" % name)
+
+        if not isinstance(data, dict):
+            return bad("'%s' is not a list" % name)
```
```suggestion
            return bad(f"{name} is not a dict")
```

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,59 @@ def bad(msg):
+        """
+        Feedback.check_dict(name, ref, data)
+
+        Check that a student dict has correct length with respect to a reference dict.  Can also check for a homogeneous data type for the list.
```
It says that only the length is checked for, but it looks like this function checks for equality between the dicts.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,59 @@ def bad(msg):
+        if ref == data:
+            return True
+        else:
+            return bad("'%s' is incorrect" % (name))
```
Having more detailed feedback would be useful, since it's not necessarily clear whether the keys or keys and values are being checked for equality to the end user.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,59 @@ def bad(msg):
+        - ``data``: Student dictionary to be checked.
+        - entry_type: If not None, checks if each value in the student dictionary has this type.
+        - ``accuracy_critical``: If true, grading will halt on failure.
+        - ``report_failure``: If true, feedback will be given on failure.
```
Some more parameters would be useful, like checking keys and values or just values only, and maybe even checking that the student answer contains some keys in particular instead of just equality.

harivamsi9 replied:
I have added:
- ``partial_keys``: If not None, it takes a List of keys to check if these particular keys are present in the student's dict or not.
- ``check_only_values``: If true, grading will be done only based on checking all valeus in student's dict and reference's dict match or not.
- ``check_only_keys``: If true, grading will be done only based on checking all keys in student's dict and reference's dict match or not.

harivamsi9 replied:
- ``entry_type_key``: If not None, requires that each key in the student's dictionary in solution be of this type.
- ``entry_type_value``: If not None, requires that each value in the student's dictionary in solution be of this type.

eliotwrobson reviewed (changes requested):
Early comments. Possibly over-engineering things a little bit but I think this function should be fairly robust to fill a couple of different use cases.

eliotwrobson commented on docs/python-grader/index.md:
```diff
@@ -164,6 +164,8 @@ The code feedback library contains built-in functions for checking correctness o
   Checks that a numpy array has the same shape and datatype as the reference solution, also checks values to see if they are close using specified `rtol` and `atol`.
 - `check_list(name, ref, data, entry_type=None)`
   Checks that a list has the same length as the reference solution. If `entry_type` is not `None`, can optionally check if each element has that type. Does _not_ check values against the reference.
+- `check_dict(name, ref, data, entry_type_key=None, entry_type_value=None)`
```
Function signature here is missing new keyword arguments.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+            if check_only_keys:
+                for key in data.keys():
+                    if key not in ref.keys():
+                        return bad(f"{name} contains an extra key: {key}")
```
No error gets reported if there is a key in ref.keys() that is not in data.keys() (i.e. extra keys). This functionality is confusing since it seems like this use case should be covered by `check_only_keys`, but is instead by `partial_keys`. I think these should be consolidated, or change `partial_keys` into a flag that disables the length equality check.

harivamsi9 replied:
@eliotwrobson thanks for pointing this out! 


The reason why I did not add additional code for "No error gets reported if there is a key in ref.keys() that is not in data.keys() (i.e. extra keys)." functionality is because, we already perform this check as a default check in the check_dict() function Line: 347

Ref: https://github.com/PrairieLearn/PrairieLearn/pull/9436/files#r1574047293

harivamsi9 replied:
Otherwise, I could change the line 347 
`        if len(ref) != len(data):
`
to 

`        if partial_keys != None and len(ref) != len(data):`

So that, when we are only checking for partial keys, we skip the the length equality check.

eliotwrobson replied:
I think that the second choice (adding the partial_keys check) makes things clearer.

eliotwrobson commented on docs/python-grader/sphinx-docs.md:
```diff
@@ -62,6 +62,30 @@ Check that a student list has correct length with respect to a reference list. C
 
+#### classmethod check_dict(name, ref, data)
+
+Checks that a student dict has all correct key-value mappings with respect to a reference dict. Can also check that a student dict has correct length of keys with respect to a reference dict's length of keys.  Can also check for a homogeneous data type for either the dict's key, dict's value, or both. Can also check if only some particular keys present in the student dict or not. Can also check s ...
```
I think this is a little confusing. The `check_only_keys` works mostly the way I would expect, but `check_only_values` checks that the set of values is equal, but ignores the keys they are mapped to. I think we should get rid of the `check_only_values` flag but keep the `check_only_values` flag (defaulting to full equality if this flag is false).

harivamsi9 replied:
I believe the current code covers this functionality.  

i.e. 
if `check_only_values` flag is True, then we solely evaluate the student's submission based on the "set of values is equal, but ignores the keys they are mapped to." 

and if `check_only_values` flag is False, then we perform the proper full equality i.e. `ref == data` Line 391.

Maybe I need to write the documentation more elaborately. 

@eliotwrobson Is this what you are looking for? or please let me know if you need any chances here.

eliotwrobson replied:
I think that makes sense. More detailed documentation would make the logic much easier to follow.

harivamsi9 replied:
Awesome! I will update the documentation to keep it unambiguous

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+            ref == data
+        ):  # will check equality of both keys and values between reference dict and student's dict
+            return True
+        else:
```
```suggestion

```
No `else:` needed, since the if case above returns (so can just return the result of calling `bad`).

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+
+        if len(ref) != len(data):
+            return bad(
+                f"{name} has the wrong length for keys--expected {len(ref)}, got {len(data)}"
```
```suggestion
                f"{name} has the wrong number of entries, expected {len(ref)}, got {len(data)}"
```

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+            )
+
+        if entry_type_value is not None:
+            for _, value in data.items():
```
```suggestion
            for value in data.values():
```

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+                    return bad(f"{name} has the wrong type for value {value}")
+
+        if entry_type_key is not None:
+            for key, _ in data.items():
```
```suggestion
            for key in data.keys():
```

eliotwrobson reviewed (commented):
More comments. Some of the functionality here has some overlapping use cases that could be confusing to end users. We're coming up with the interface for this on the fly, so not a huge deal if it takes a couple rounds of review to iron things out.

eliotwrobson commented:
@harivamsi9 do you think you'll be able to see this through? Sorry if the ambiguity from me is slowing things down here, if things have gotten a little confusing I'd be happy to post on slack to get some clearer direction on what use cases others may want.

harivamsi9 commented:
Hey Sorry for the delay, I was procrastinating things! I will wrap this up by tomorrow! 
Happy to chat about any additional changes required over slack.

eliotwrobson commented:
@harivamsi9 any updates here? If you're too busy, I'm happy to take over this PR, but it looks like it's in a pretty close state to being done. We'd like to have this done before the end of the month if possible 🙏🏽

harivamsi9 commented:
hey @eliotwrobson i will finish it off

harivamsi9 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,103 @@ def bad(msg):
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if len(ref) != len(data):
```
Line: 347
Here is where we check "No error gets reported if there is a key in ref.keys() that is not in data.keys() (i.e. extra keys)."

harivamsi9 commented:
Hey @eliotwrobson, I believe I have covered all the review items, and also added more detailed documentation. 
I do not know why the `CI/native-checks (pull_request)` is failing. Please let me know what can I do to rectify this. 

Apart from that, I believe this PR is good to go. Could you please review this and merge if not issues found?

eliotwrobson commented:
@harivamsi9 will take a look soon! The native checks are failing because of the formatting in the documentation. You have to run prettier over all of the markdown files you change (this can be done by the `make format` command in the directory https://prairielearn.readthedocs.io/en/latest/dev-guide/#coding-style).

harivamsi9 commented:
> @harivamsi9 will take a look soon! The native checks are failing because of the formatting in the documentation. You have to run prettier over all of the markdown files you change (this can be done by the `make format` command in the directory https://prairielearn.readthedocs.io/en/latest/dev-guide/#coding-style).

When I try to fix this, I encounter this error

<img width="1445" alt="image" src="https://github.com/PrairieLearn/PrairieLearn/assets/44733128/c5645806-6a4b-410c-90e1-088e4ea18fa7">

nwalters512 commented:
You'll need to run `make deps` first. In general, you should follow our documentation for setting up your dev environment. Pick the appropriate one depending on if you're running natively or in Docker:

- https://prairielearn.readthedocs.io/en/latest/installingLocal/
- https://prairielearn.readthedocs.io/en/latest/installingNative/

harivamsi9 commented:
> You'll need to run `make deps` first. In general, you should follow our documentation for setting up your dev environment. Pick the appropriate one depending on if you're running natively or in Docker:
> 
> * https://prairielearn.readthedocs.io/en/latest/installingLocal/
> * https://prairielearn.readthedocs.io/en/latest/installingNative/

While replicating the steps given for docker, when I enter `make deps`, I encounter the following error. 
<img width="1220" alt="image" src="https://github.com/PrairieLearn/PrairieLearn/assets/44733128/d0ae0d1b-dd85-4c23-91ac-9e21afd70e41">

I tried to resolve this, yet did not find any solution. So I went ahead with the `make dev` command which gave me this error 
<img width="1352" alt="image" src="https://github.com/PrairieLearn/PrairieLearn/assets/44733128/14eb42dd-2785-42db-86a1-699187f11b1c">

Please let me know if there is something I am doing wrong here. 

Meanwhile, I am retrying with https://prairielearn.readthedocs.io/en/latest/installingNative/ approach...

harivamsi9 commented:
Also, just an another idea, 
If we know that this error usually comes up when formatting in the documentation, can we perform this `make format` command in the CI itself? or may be using something like a git hooks perhaps. Please correct me if I am wrong here

nwalters512 commented:
I see `KeyboardInterupt` in your error message, did you press Ctrl/Cmd-C while running? Did you try running `make deps` again? If that fails, other things won't work either.

We generally don't want CI to mutate branches, and because of how many folks tend to do PrairieLearn development (inside a Docker container), Git hooks wouldn't work well because the necessary dependencies wouldn't be installed in a place that's accessible to any such hooks.

harivamsi9 commented:
> I see `KeyboardInterupt` in your error message, did you press Ctrl/Cmd-C while running? Did you try running `make deps` again? If that fails, other things won't work either.
> 
> We generally don't want CI to mutate branches, and because of how many folks tend to do PrairieLearn development (inside a Docker container), Git hooks wouldn't work well because the necessary dependencies wouldn't be installed in a place that's accessible to any such hooks.

Ohh I see. 

It is kinda strange, because I did not press cmd/ctrl+c or any other keyboard interrupts. I used both the latest PL pull and as well as the with my branch. 


Update: Finally, ALL GREENS AT LAST, was able to setup everything properly when I used the 2nd approach (i.e. InstallingNative), and make the `make format` work. Also, the problem still persists on the docker approach (i.e. installingLocal). 

I would like to point out that the "Install the prerequisites:" sections needs to be updated to include R software as well. There was a problem when running `make deps` which will invoke pip/pip3 install rpy2, but this throws up an error like 
<img width="1454" alt="image" src="https://github.com/PrairieLearn/PrairieLearn/assets/44733128/8ba9feab-ae35-4c49-8f81-9800d99a8bb4">

Upon debugging further, I found out that R was required to be installed (reference: https://stackoverflow.com/questions/17573988/r-home-error-with-rpy2). So I believe it will be helpful to add R also into the "Install the prerequisites:", so it is a bit easier for the first-timers to on-board. 

That being said, this is my First proper PR to PraireLearn, from being a Teaching Assistant for CSE 8A: Intro to Programming at UCSD to using PraireLearn to code/build/administrate our exams, quizes, homeworks to identifying small gaps in the system and proposing ideas, and actually contributing to the software.  

As a few next steps, I am planning on creating content for YT, Insta, Twitter(X), LinkedIn to talk about this PR, and also to help people start with OSS and also to build in public. I deeply appreciate @eliotwrobson and @nwalters512 support throughout this, and I am interested to keep contributing to develop new features, take up bigger tasks, write high quality code, and scale PraireLearn.

eliotwrobson commented on exampleCourse/questions/demo/annotated/MarkovChainGroupActivity/MarkovChains-PageRank/serverFilesQuestion/Markov-Chains-3.ipynb:
It looks like a bunch of unrelated changes to the jupyter notebooks got committed? We'll need to remove before merging.

harivamsi9 replied:
yes, these jupyter notebooks are modified with the `make format` cmd. I assumed that some of the previous merges might have missed this, so I committed even those. 
I will remove them, and keep this PR only specific to my task.

harivamsi9 replied:
I am familiar with using `pytest`. I need to progress on my other project, I will get back to working on this tomorrow, and connect with you over slack if I needed any further clarifications.

nwalters512 commented:
Huh, I really don't know what's happening when you try to execute in Docker. If you want to get that working, you could start a GitHub Discussion to look at that in more detail (PR comments aren't the best venue to debug dev setup). That error doesn't look like anything I've seen before, and if there's a possibility it could impact other people, it'd be nice to figure out what's going on.

R is a weird one, we definitely want better guidance there. Installing R is of course on option, but since it's being phased out, you can also just temporarily comment out `rpy2` from the requirements file before running `make deps`. Sorry if you wasted time on that!

We're glad to have you here contributing 🙌

eliotwrobson reviewed (changes requested):
A bit of an incomplete review, but I wanted to give one piece of feedback early. Do you think you could write a couple of test cases for this new function? The Python autograder doesn't have a ton of dev tools set up, so it would be much easier to verify correctness of the changes with some test cases.

We're using `pytest` for these. It should be possible to add a new test file by looking at the `pyproject.toml` file in the root directory. You can add the new test file to a new test-subdirectory in the python autograder folder. I know these are somewhat vague instructions, so _please_ follow up with me over DM on slack if you have trouble setting up this new test file.

nwalters512 commented:
@harivamsi9 Eliot will be largely unavailable this summer, so I'll be the point of contact / reviewer going forward!

nwalters512 commented:
@harivamsi9 are you still interested in taking this over the finish line? If not, I'll assign this to someone else. Thanks!

harivamsi9 commented:
Hey @nwalters512, sorry for the delay. Let me do a quick sprint and bring this to the finish line ASAP. Thanks for circling back on this.

harivamsi9 commented:
Hi @eliotwrobson 
Hope you are doing well. 

I have drafted the testcases for the check_dict(), and tested the functionality under various scenarios. 
Here is the coverage report from pytest:
![image](https://github.com/user-attachments/assets/11ffb049-a1bf-43ae-aa83-a0f1f61a2b85)

- added the tests/ sub-dir with tests for check_dict()
- added __init__.py files to /python_autograder/tests/  and /python_autograder/ to make the import of code_feedback.py to the test file
- Achieved code coverage of 31% (previously 0%) on code_feedback.py and overall 64%
- Added 38 test custom test cases to check and valid the check_dict() thoroughly under different scenarios. 


There are a few issues I see in here, which I would like to ask you: 

Should I add the test file to a new sub-directory in python_autograder as you suggested? 
Like python_autograder/tests/code_feedback_test.py                       ---- 1
Or
should I just add the test file directly to python_autograder folder? ---- 2

The problem in each is that, 
1. tests/ sub-dir method:

- we need add a new file __ init__.py inside both tests/  and python _autograder/ 
- This is to make the code_feedback module importable from /tests/
- Pros - tested with pytest on the tests/ dir and passing all testcases. and you see the output as shown in the above pytest coverage report. 
used this cmd: `pytest graders/python/python_autograder/tests`

- Cons - need to rename all module imports in all python files inside the python_autograder 
from “import code_feedback” to “import  python_autograder.code_feedback” and this should be done to other modules as well. 
Like, say pl_result.py imports the pl_execute, code_feedback, pl_helpers modules, 
now all of the imports should be updated to python_autograder.pl_execute,  python_autograder.code_feedback, 
 python_autograder.pl_helpers respectively. 
 
2. If directly added the test file to the python_autograder, then 
When i run pytest, it fails to execute because of some import issue of nbformat 
Even after i did the pip3 install nbformat


I hope I was able to give you some context from above example. Please let me know what you think is the best way to do this?

Also, after the recent tests update, the native-checks are failing, any idea where I should be looking to resolve this?

eliotwrobson commented on graders/python/python_autograder/tests/code_feedback_test.py:
Haven't looked at every test case, but a lot of these test cases seem to be very similar. I think many of these can be condensed by parameterizing these test cases: https://docs.pytest.org/en/stable/example/parametrize.html

Also, should add type annotations to this file.

eliotwrobson commented on pyproject.toml:
```diff
@@ -95,6 +95,8 @@ include = [
     "./apps/prairielearn/python/zygote.py",
     "./apps/prairielearn/python/zygote_utils.py",
     "./apps/prairielearn/python/zygote_utils_test.py",
+    "./graders/python/python_autograder/code_feedback.py",
```
No need to include the non-test autograder files, since they don't have type annotations and we probably won't add them in the future.

eliotwrobson commented:
@harivamsi9 to your question, I think that having a tests subdirectory is the way to go, this makes it easier to view the files in the autograder. I'll have to double check, but I think adding `__init__.py` files to the autograder will break existing autograder code? There's probably a way to avoid adding these files but still import what you want for the test cases using importlib: https://docs.python.org/3/library/importlib.html

As for the failing native checks, see my other comments. This error may resolve itself after making the changes I mentioned there.

harivamsi9 commented:
Hey @eliotwrobson 
I will work on this today. I see that you have committed some changes. please let me know if any activities are pending.
 I can take care of them

eliotwrobson commented:
@harivamsi9 I think everything is looking ok! I managed to get everything running, so just take a look and see if the test cases are comprehensive. It might also be good to add a question to the example course that uses this method in the grading algorithm, but given the comprehensive test cases I don't think that's strictly necessary.

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+
+            return check_keys or check_values
+
+        if (
+            ref == data
+        ):  # will check equality of both keys and values between reference dict and student's dict
+            return True
```
Nit for better formatting:

```suggestion
        # Check equality of both keys and values between reference dict and student's dict
        if ref == data:  
            return True
```

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+        ):  # will check equality of both keys and values between reference dict and student's dict
+            return True
+
+        return bad(f"{name} is incorrect as one (or more) key-value pairs do not match")
```
```suggestion
        return bad(f"{name} has one or more key-value pairs that do not match")
```

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+            check_values = False
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return f"{name} has the wrong length for values--expected {len(ref.values())}, got {len(data.values())}"
```
```suggestion
                    return f"{name} has the wrong length for values: expected {len(ref.values())}, got {len(data.values())}"
```

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if partial_keys is not None and len(partial_keys) >= 1:
+            for partial_key in partial_keys:
+                if partial_key not in data:
+                    return bad(f"{name} does not contain key {partial_key}")
+            return True
```
The behavior here is unintuitive to me: if `partial_keys` is specified, none of the other checks will run, so something like `check_dict(..., partial_keys=["a"], entry_type_value=int)` won't do what it looks like it will.

harivamsi9 replied:
Yes! I see what you are saying. I somehow missed this piece. I have modified the code to reflect the logic to handle the combos of different parameters. Please let me know in case any changes are required.

eliotwrobson replied:
I think this has since been removed? Going to resolve this item.

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+        - ``partial_keys``: If not None, it takes a List of keys to check if these particular keys are present in the student's dict or not.
+        - ``check_only_keys``: If true, grading will be done only based on checking all keys in student's dict and reference's dict match or not.
+        - ``check_only_values``: If true, grading will be done only based on checking all values in student's dict and reference's dict match or not.
+        - ``entry_type_key``: If not None, requires that each key in the student's dictionary in solution be of this type.
+        - ``entry_type_value``: If not None, requires that each value in the student's dictionary in solution be of this type.
```
Why `entry_type_key` and `entry_type_value` instead of `value_type` and `key_type`?

harivamsi9 replied:
Yes! My bad! We could simply use `value_type` and `key_type`.

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,112 @@ def bad(msg):
+                    )
+
+        if check_only_keys or check_only_values:
+            check_keys = False
```
I'm trying to tell what's going on here. Would this be better named `keys_valid`, `keys_checked`, `checked_keys`, or something like that? Or am I misunderstanding what this variable represents?

eliotwrobson replied:
I think these are confusing variable names. More expressive would be `check_keys` and `check_values`, defaulted to True. Then `check_keys` would require having `check_values` set to True. Let me think about this some more.

harivamsi9 replied:
Let me look into all the mentioned comments right now. And I will clarify them shortly.

harivamsi9 replied:
Lines 379 - 399

`check_only_keys` - When True, the grading will only consider if student's dict and reference's dict have the same `keys` or not. Once this comparison is made, the function has to return True if matched, False otherwise. 

`check_only_values` - When True, the grading will only consider if student's dict and reference's dict have the same `values` or not. Once this comparison is made, the function has to return True if matched, False otherwise. 

As you can see, `check_only_keys` and `check_only_values` are independent checks meaning as a user, I can just want to enable grading based on only values and choose to give credit, or based on only keys and just give credit, or choose to have both. In case, both are enabled then the logic should confirm that both these checks return True, and only then check_dict() can return true. ( This is handled in lines: 396, 397)

At the same time, if only one of these is enabled, they logic should be intact and return True if the specific check returns true.  ( This is handled in lines: 399)

And the `check_keys` and `check_values` are local flags helping us to build the above logic.


Is this what you are looking for? Please let me know your thoughts.

harivamsi9 replied:
Actually, I think I understand your question now. 

A better naming would be to use: 
`check_keys` instead of `check_only_keys` 
`check_values` instead of `check_only_values`  
`keys_valid` instead of `check_keys` 
`values_valid` instead of `check_values` 

Is this what you are suggesting?

eliotwrobson replied:
^I think this simpler naming would be helpful.

eliotwrobson replied:
Latest push includes these renamed variables 👍🏽

eliotwrobson commented on docs/python-grader/index.md:
```diff
@@ -164,6 +164,8 @@ The code feedback library contains built-in functions for checking correctness o
 - `check_list(name, ref, data, entry_type=None)`
   Checks that a list has the same length as the reference solution. If `entry_type` is not `None`, can optionally check if each element has that type. Does _not_ check values against the reference.
+- `check_dict(name, ref, data, partial_keys=None, check_only_keys=False, check_only_values=False, entry_type_key=None, entry_type_value=None, accuracy_critical=False, report_failure=True)`
+  Checks that a student dict has all correct key-value mappings with respect to a reference dict. Can also check that a student dict has correct length of keys with respect to a reference dict's length of keys. Can also check for a homogeneous data type for either the dict's key, dict's value, or both. If `entry_type_key` is not `None`, can optionally check if each key in the student's dictionary ...
```
It checks values if the corresponding flags are set. Let me run through this function and clear up some of the documentation.

eliotwrobson replied:
@harivamsi9 could you clarify the documentation here a little bit?

eliotwrobson replied:
@nwalters512 the last sentence was accidentally copied from the documentation of another method, I've removed it in the latest push.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -299,6 +299,122 @@ def bad(msg):
+                        return bad(f"{name} contains an extra value: {value}")
+                check_values = True
+
+            if check_only_keys and check_only_values and check_partial_keys:
```
Is there a simpler way to write the if-cases of this function below? There should be a way of doing that avoids checking every single case.

eliotwrobson replied:
I updated this in the latest push

eliotwrobson reviewed (changes requested):
More comments

harivamsi9 commented:
Hey @eliotwrobson Please let me know if any changes are needed for this PR.

eliotwrobson commented:
@harivamsi9 thanks for checking in! After some discussion, I think we just want to make sure the current API for the feature in this PR is flexible enough to handle a few different use cases. Do you happen to have any examples of questions that use this, and if so, could you include them in the test suite? The code itself looks pretty solid.

reteps commented on docs/python-grader/index.md:
```diff
@@ -179,6 +179,8 @@ The code feedback library contains built-in functions for checking correctness o
   Checks that a numpy array has the same shape and datatype as the reference solution, also checks values to see if they are close using specified `rtol` and `atol`.
 - `check_list(name, ref, data, entry_type=None)`
   Checks that a list has the same length as the reference solution. If `entry_type` is not `None`, can optionally check if each element has that type. Does _not_ check values against the reference.
+- `check_dict(name, ref, data, partial_keys=None, check_only_keys=False, check_only_values=False, entry_type_key=None, entry_type_value=None, accuracy_critical=False, report_failure=True)`
```
In keeping with #11107 , potentially we could rename this `check_strategy="keys"|"values"|"all"`

eliotwrobson replied:
That's a good idea actually. Would it make sense to use a string or an enum? I think the enum is a little cleaner but not sure if it causes issues with the way the grader is laid out.

reteps replied:
I prefer a string with exhaustiveness checking. Enums require imports.

eliotwrobson replied:
I think that's reasonable for something used by client code 👍

reteps commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -304,6 +304,110 @@ def bad(msg):
+        - ``name``: Name of the dictionary that is being checked. This will be used to give feedback.
+        - ``ref``: Reference dictionary.
+        - ``data``: Student dictionary to be checked.
+        - ``partial_keys``: If not None, it takes a List of keys to check if these particular keys are present in the student's dict or not.
```
Not a fan of the name `partial`. I would prefer `only_keys`.

eliotwrobson replied:
Maybe `required_keys`?

reteps replied:
I kept it as `only_keys`, but I would be happy to workshop the wording.

reteps commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -304,6 +304,110 @@ def bad(msg):
+        """
+        Feedback.check_dict(name, ref, data)
+
+        Checks that a student dict (`data`) has all correct key-value mappings with respect to a reference dict (`ref`). It also verifies the length of keys in the student dictionary against the reference dictionary, and optionally, enforces homogeneous data types for keys (using `key_type`), values (using `value_type`), or both. Additionally, it can verify the presence of specific keys (using `p ...
```
This comment is out of date, with regards to `check_only_keys` etc. I hope we can use my suggestion above with `check_strategy` to avoid all these boolean flags.

reteps replied:
This is outdated, all keys and values are now checked.

reteps commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -304,6 +304,110 @@ def bad(msg):
+        partial_keys=None,
+        check_keys=False,
+        check_values=False,
+        key_type=None,
```
`key_type` and `value_type` don't feel that useful to me. Why not just have the types related to the `ref` type? E.g. `type(ref["foo"]) == type(data["foo"])`

reteps replied:
I would prefer removing them and replacing them with a different name; e.g. `typecheck = "keys" | "values" | "all" | "none"`, which matches against ref.

eliotwrobson replied:
I think using an enum makes things a lot cleaner here as well.

reteps replied:
I'm thinking that all values will be typechecked, and that this doesn't need to be configured.

reteps replied:
I'm also not even sure how key type checking is useful:
```
>>> {1:2}.keys()
dict_keys([1])
>>> {'1':2}.keys()
dict_keys(['1'])
```

reteps commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -304,6 +304,110 @@ def bad(msg):
+        check_values=False,
+        key_type=None,
+        value_type=None,
+        accuracy_critical=False,
```
I don't think the name `accuracy_critical` describes the flag accurately; perhaps `abort_invalid=True|False`? I'm not a fan of that new name much better, maybe @eliotwrobson you have another suggestion?

reteps replied:
It looks like this is a convention from other functions; so I will resolve this.

reteps commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -304,6 +304,110 @@ def bad(msg):
+        if ref == data:
+            return True
+
+        return bad(f"{name} has one or more key-value pairs that do not match")
```
It might be nice to have options for more verbose error messages (key X did not match)

eliotwrobson replied:
I think that would be very useful in many situations 👍

reteps replied:
Fixed.

reteps reviewed (changes requested):
There are some API decisions I want to confirm before this is merged.

reteps commented:
@eliotwrobson I simplified the API in [010645d](https://github.com/PrairieLearn/PrairieLearn/pull/9436/commits/010645dcb22d23a442be340f0c67a1c55c0215a2). I felt that the current one includes a variety of parameters that wouldn't be used commonly. All that is remaining is to fixup the test suite.

eliotwrobson commented:
@reteps thanks! I'm a bit busy today but can make changes / review later tonight, feel free to make changes in the meantime (it looks like the typechecking is failing for some reason?).

reteps commented:
> @reteps thanks! I'm a bit busy today but can make changes / review later tonight, feel free to make changes in the meantime (it looks like the typechecking is failing for some reason?).

This is because the test suite needs to be rewritten. I also wasn't really a fan of the original way it was written.

reteps commented on graders/python/python_autograder/tests/code_feedback_test.py:
```diff
@@ -0,0 +1,277 @@
+    ],
+)
+@patch("code_feedback.Feedback.add_feedback")
+@patch("builtins.open", new_callable=mock_open)
```
We need to patch `open` so that we don't write any files.

reteps commented on graders/python/python_autograder/tests/code_feedback_test.py:
```diff
@@ -0,0 +1,277 @@
+# ruff: noqa: ARG001 ANN001
+# pyright: reportUnknownParameterType=none, reportMissingParameterType=none
```
I didn't bother typing the `Mock` objects.

reteps commented:
Docs: https://prairielearn--9436.org.readthedocs.build/en/9436/python-grader/reference-docs/#code_feedback.Feedback.check_dict

reteps commented:
Still need to add an example question using this.

eliotwrobson commented on docs/python-grader/index.md:
```diff
@@ -191,6 +191,8 @@ The code feedback library contains built-in functions for checking correctness o
   Checks that a numpy array has the same shape and datatype as the reference solution, also checks values to see if they are close using specified `rtol` and `atol`.
 - `check_list(name, ref, data, entry_type=None)`
   Checks that a list has the same length as the reference solution. If `entry_type` is not `None`, can optionally check if each element has that type. Does _not_ check values against the reference.
+- `check_dict(name, ref, data, *, partial_keys=None, check_keys=False, check_values=False, key_type=None, value_type=None, accuracy_critical=False, report_failure=False)`
```
This has parameters that were removed.

reteps replied:
Fixed in b0154849a9187cec010fdefb137a06043532db88.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -334,6 +334,81 @@ def bad(msg: str) -> Literal[False]:
+                return False
+
+        if data is None:
+            return bad(f"{name} is not a dict, got None")
```
This seems a little redundant with the case below? I recognize the error message is a little nicer, but I don't really see this  specific case being a common error.

reteps replied:
Fixed in b0154849a9187cec010fdefb137a06043532db88.

eliotwrobson commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -334,6 +334,81 @@ def bad(msg: str) -> Literal[False]:
+        ref: dict[Any, Any],
+        data: dict[Any, Any],
+        *,
+        only_keys: None | list[str] = None,
```
Maybe `target_keys` is a better variable name? The current name makes it seem like this should be a boolean. I also think this should be a set since we don't want duplicates.

reteps replied:
Fixed in b0154849a9187cec010fdefb137a06043532db88.

reteps replied:
We would have to do `list -> set -> sort -> list` to support any sort of deterministic ordering on the keys.

reteps replied:
In [fa3fde1](https://github.com/PrairieLearn/PrairieLearn/pull/9436/commits/fa3fde1fd0dd6ff3c46b47d15e195e4edd3984b1) I support a set. No code changes needed.

eliotwrobson reviewed (approved):
Simplified interface seems a lot more natural and new test cases seem pretty comprehensive. A couple of small comments but everything else looks good. Not sure if we should worry about having an example question right away, it probably makes more sense to add that in after the fact, and the simpler interface makes this straightforward to use without an example.

reteps commented:
I'll wait for @nwalters512 to make a final decision on if the example question is needed. IMO, the comprehensive test coverage should be enough.

nwalters512 commented on docs/python-grader/index.md:
```diff
@@ -191,6 +191,8 @@ The code feedback library contains built-in functions for checking correctness o
   Checks that a numpy array has the same shape and datatype as the reference solution, also checks values to see if they are close using specified `rtol` and `atol`.
 - `check_list(name, ref, data, entry_type=None)`
   Checks that a list has the same length as the reference solution. If `entry_type` is not `None`, can optionally check if each element has that type. Does _not_ check values against the reference.
+- `check_dict(name, ref, data, target_keys=None)`
+  Checks that a student dict has all correct key-value mappings with respect to a reference dict.
 - `check_tuple(name, ref, data)`
```
Half of those options don't exist. Look again; are you maybe looking at an outdated revision of the code?

nwalters512 reviewed (approved):
Agreed that with the tests + docs, an example question isn't strictly necessary. That said, I'll assume that you've also manually tested the final container image on a real question and confirmed that it's working as well as the tests say they do 🙂

nwalters512 commented on graders/python/python_autograder/tests/code_feedback_test.py:
```diff
@@ -0,0 +1,272 @@
+
+
+@pytest.mark.parametrize(
+    ("test_id", "ref_dict", "student_dict"),
```
Idk if it's just the Pythonic way but I find comments at the end of lines to be much more annoying to mentally parse than a comment that precedes the relevant lines. If that's just me, I'm happy to ignore it.

nwalters512 commented on graders/python/python_autograder/code_feedback.py:
```diff
@@ -334,6 +334,83 @@ def bad(msg: str) -> Literal[False]:
+                )
+
+        if report_success:
+            cls.add_feedback(f"{name} looks good")
```
The addition of this made me go and look at other implementation of `report_success`, where I noticed that they all did the following:

```suggestion
            cls.add_feedback(f"'{name}' looks good")
```

Consider making this change for the other feedback as well.

reteps replied:
Done in [5140263](https://github.com/PrairieLearn/PrairieLearn/pull/9436/commits/51402635b285c6e458545ddbd3ba175a912089b6).

reteps commented:
Tested a python-autograder question and ensured that it was still working.
